File: algorithms/[39]组合总和/solution.py
# topics = ["动态规划"]

from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        """
        近似题 518. 零钱兑换 II
        完全背包 & 恰好装满，求所有方案（组合）
        """
        # dp[i] 表示组成目标数 i 所有的组合
        dp: List[List[List[int]]] = [[] for _ in range(target + 1)]
        dp[0] = [[]]

        # 先遍历物品， 再遍历背包大小！
        for c in candidates:
            for i in range(1, target + 1):
                if i >= c:
                    dp[i].extend([x + [c] for x in dp[i - c]])

        return dp[target]

File: algorithms/[39]组合总和/solution.ts
// topics = ["回溯算法"]

/**
 * 回溯法搜索所有可行解模板
 * res = []
 * path = []
 *
 * def backtrack(未探索区域, res, path):
 *     if path 满足条件:
 *         res.add(path)  # 深度拷贝
 *         # return       # 如果不用继续搜索需要 return
 *     for 选择 in 未探索区域当前可能的选择:
 *         if 当前选择符合要求:
 *             path.add(当前选择)
 *             backtrack(新的未探索区域, res, path)
 *             path.pop()
 */

function combinationSum(candidates: number[], target: number): number[][] {
  const n = candidates.length;
  const res: number[][] = [];
  const path: number[] = [];

  const backtrack = (idx = 0): void => {
    const sum = path.reduce((acc, cur) => acc + cur, 0);
    if (sum == target) {
      res.push([...path]);
      return;
    }
    for (let i = idx; i < n; i += 1) {
      if (sum + candidates[i] <= target) {
        path.push(candidates[i]);
        backtrack(i);
        path.pop();
      }
    }
  };

  backtrack();

  return res;
}
